src/tz.cpp

```diff
@@ -267,6 +294,83 @@ get_download_folder()
 
 #  endif  // !_WIN32
 
+
+/*
+ * This class is provided to mimic the following usage of `ifstream`:
+ *
+ * std::ifstream is(filename);
+ *
+ * file_streambuf ibuf(filename);
+ * std::istream is(&ibuf);
+ *
+ * This is required because `ifstream` does not support opening files
+ * containing wide characters on Windows. On Windows, `file_streambuf` uses
+ * `file_open()` to convert the file name to UTF-16 before opening it with
+ * `_wfopen()`.
+ *
+ * Note that this is not an exact re-implementation of `ifstream`,
+ * but is enough for our usage.
+ *
+ * It is partially based on these two implementations:
+ * - fdinbuf from http://www.josuttis.com/cppcode/fdstream.html
+ * - stdiobuf https://stackoverflow.com/questions/12342542/convert-file-to-ifstream-c-android-ndk
+ *
+ * Apparently MSVC provides non-standard overloads of `ifstream` that support
+ * a `const wchar_t*` file name, but MinGW does not.
+ * https://stackoverflow.com/a/822032
+ */
+class file_streambuf
+  : public std::streambuf
+{
+private:
+    FILE* file_;
+    static const int buffer_size_ = 1024;
+    char buffer_[buffer_size_];
+
+public:
+    file_streambuf(const std::string& filename)
+        : file_(file_open(filename))
+    {
+    }
+
+    ~file_streambuf()
+    {
+        if (file_) {
+            ::fclose(file_);
+        }
+    }
+
+protected:
+    virtual int_type underflow() {
+        if (gptr() == egptr() && file_) {
+            const size_t size = ::fread(buffer_, 1, buffer_size_, file_);
+            setg(buffer_, buffer_, buffer_ + size);
+        }
+        return (gptr() == egptr())
+            ? traits_type::eof()
+                : traits_type::to_int_type(*gptr());
+    }
+
+private:
+    FILE*
+    file_open(const std::string& filename)
+    {
+#  ifdef _WIN32
+        std::wstring wfilename = to_utf16(filename);
+        FILE* file = ::_wfopen(wfilename.c_str(), L"r");
+#  else // !_WIN32
+        FILE* file = ::fopen(filename.c_str(), "r");
+#  endif // _WIN32
+        if (file == NULL) {
+            std::string msg = "Error opening file \"";
+            msg += filename;
+            msg += "\".";
+            throw std::runtime_error(msg);
```

Comment:
Not directly related to this PR, but just throwing a generic error on a file error is not great, because you throw away the original error code and error message. Opening a file might fail for many different reasons. So ideally you would check `errno` on error, or use the C++ equivalent (which might need C++11, but I guess that's fine).

Reply:
I'll probably try and see if this code could be accepted into date itself, and based on this comment they haven't been caring much about the specific reason for the failures https://github.com/r-lib/tzdb/pull/13/files#diff-ddeec8b3654d088baf9401ba973b87219e7f27f2b1e7f37a0062014cf28143e5L565

But that is great to know for the future!